File: jambeth/jambeth-cache-datachange/src/main/java/de/osthus/ambeth/cache/RootCacheClearEventListener.java
package de.osthus.ambeth.cache;

import de.osthus.ambeth.collections.IList;
import de.osthus.ambeth.event.IEventListener;
import de.osthus.ambeth.ioc.IInitializingBean;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.util.ParamChecker;

public class RootCacheClearEventListener implements IEventListener, IInitializingBean
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	protected IFirstLevelCacheManager firstLevelCacheManager;

	protected ISecondLevelCacheManager secondLevelCacheManager;

	protected IRootCache committedRootCache;

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		ParamChecker.assertNotNull(firstLevelCacheManager, "FirstLevelCacheManager");
		ParamChecker.assertNotNull(secondLevelCacheManager, "SecondLevelCacheManager");
		ParamChecker.assertNotNull(committedRootCache, "CommittedRootCache");
	}

	public void setFirstLevelCacheManager(IFirstLevelCacheManager firstLevelCacheManager)
	{
		this.firstLevelCacheManager = firstLevelCacheManager;
	}

	public void setSecondLevelCacheManager(ISecondLevelCacheManager secondLevelCacheManager)
	{
		this.secondLevelCacheManager = secondLevelCacheManager;
	}

	public void setCommittedRootCache(IRootCache committedRootCache)
	{
		this.committedRootCache = committedRootCache;
	}

	@Override
	public void handleEvent(Object eventObject, long dispatchTime, long sequenceId)
	{
		if (!(eventObject instanceof ClearAllCachesEvent))
		{
			return;
		}
		committedRootCache.clear();
		IRootCache privilegedSecondLevelCache = secondLevelCacheManager.selectPrivilegedSecondLevelCache(false);
		if (privilegedSecondLevelCache != null && privilegedSecondLevelCache != committedRootCache)
		{
			privilegedSecondLevelCache.clear();
		}
		IRootCache nonPrivilegedSecondLevelCache = secondLevelCacheManager.selectNonPrivilegedSecondLevelCache(false);
		if (nonPrivilegedSecondLevelCache != null && nonPrivilegedSecondLevelCache != committedRootCache
				&& nonPrivilegedSecondLevelCache != privilegedSecondLevelCache)
		{
			nonPrivilegedSecondLevelCache.clear();
		}
		IList<IWritableCache> firstLevelCaches = firstLevelCacheManager.selectFirstLevelCaches();
		for (int a = firstLevelCaches.size(); a-- > 0;)
		{
			IWritableCache firstLevelCache = firstLevelCaches.get(a);
			firstLevelCache.clear();
		}
	}
}

File: ambeth/Ambeth.CacheDataChange/ambeth/cache/RootCacheClearEventListener.cs
using System;
using De.Osthus.Ambeth.Event;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Util;
using System.Collections.Generic;
using De.Osthus.Ambeth.Merge.Model;

namespace De.Osthus.Ambeth.Cache
{
    public class RootCacheClearEventListener : IEventListener, IInitializingBean
    {
        [LogInstance]
	    public ILogger Log { private get; set; }

        public IFirstLevelCacheManager FirstLevelCacheManager { protected get; set; }

        public ISecondLevelCacheManager SecondLevelCacheManager { protected get; set; }

        public IRootCache CommittedRootCache { protected get; set; }

	    public void AfterPropertiesSet()
	    {
            ParamChecker.AssertNotNull(FirstLevelCacheManager, "FirstLevelCacheManager");
            ParamChecker.AssertNotNull(SecondLevelCacheManager, "SecondLevelCacheManager");
            ParamChecker.AssertNotNull(CommittedRootCache, "CommittedRootCache");
        }

	    public void HandleEvent(Object eventObject, DateTime dispatchTime, long sequenceId)
	    {
            if (!(eventObject is ClearAllCachesEvent))
		    {
			    return;
		    }
            CommittedRootCache.Clear();
            IRootCache privilegedSecondLevelCache = SecondLevelCacheManager.SelectPrivilegedSecondLevelCache(false);
            if (privilegedSecondLevelCache != null && !Object.ReferenceEquals(privilegedSecondLevelCache, CommittedRootCache))
            {
                privilegedSecondLevelCache.Clear();
            }
            IRootCache nonPrivilegedSecondLevelCache = SecondLevelCacheManager.SelectNonPrivilegedSecondLevelCache(false);
            if (nonPrivilegedSecondLevelCache != null && !Object.ReferenceEquals(nonPrivilegedSecondLevelCache, CommittedRootCache)
                    && !Object.ReferenceEquals(nonPrivilegedSecondLevelCache, privilegedSecondLevelCache))
            {
                nonPrivilegedSecondLevelCache.Clear();
            }
            IList<IWritableCache> firstLevelCaches = FirstLevelCacheManager.SelectFirstLevelCaches();
            for (int a = firstLevelCaches.Count; a-- > 0; )
            {
                IWritableCache firstLevelCache = firstLevelCaches[a];
                firstLevelCache.Clear();
            }
	    }
    }
}
